OpenRA.Mods.Common/Traits/BotModules/ResourceMapBotModule.cs

```diff
@@ -0,0 +1,313 @@
+
+					if (Info.HarvesterTypes.Contains(a.Info.Name) && a.Owner == player)
+						harvesterCount++;
+
+					return Info.ResourceCreatorTypes.Contains(a.Info.Name);
+				}).Select(a => a.Location).ToList();
+
+			indice.ResourceCellsCount = resTilesCount;
+			indice.ResourceCellsCenter = bestCell;
+			indice.ResourceCreatorLocs = resourceCreatorLocs;
+			indice.PlayerRefineryCount = refineryCount;
+			indice.PlayerHarvetserCount = harvesterCount;
+			indice.EnemyUnitCount = normalEnemyCount;
+			indice.EnemyBaseCount = highThreatEnemyCount;
+		}
+
+		public int GetIndicesLength()
+		{
+			return resourceMapIndices?.Length ?? 0;
+		}
+
+		public int GetIndiceSideLength()
+		{
+			return indiceSideLength;
+		}
+
+		public int GetIndiceColumnCount()
+		{
+			return resourceMapIndicesColumnCount;
+		}
+
+		public int GetIndiceRowCount()
+		{
+			return resourceMapIndicesRowCount;
+		}
+
+		public int GetIndiceScanRadius()
+		{
+			return indiceResourceScanRadius;
+		}
+
+		public (int IndiceCount, int EnemyUnitCount, int EnemyBaseCount) GetNearbyIndicesThreat(int index)
+		{
+			var baseIndice = resourceMapIndices[index];
+
+			var indiceCount = 0;
+			var nearbyEnemyBase = 0;
+			var nearbyEnemy = 0;
+
+			var x = baseIndice.IndiceIndexs.X;
+			var y = baseIndice.IndiceIndexs.Y;
+
+			var leftIndex = x - 1 + y * resourceMapIndicesColumnCount;
+			if (leftIndex >= 0)
+			{
+				var indice = resourceMapIndices[leftIndex];
+				nearbyEnemyBase += indice.EnemyBaseCount;
+				nearbyEnemy += indice.EnemyUnitCount;
+				indiceCount++;
+			}
+
+			var rightIndex = x + 1 + y * resourceMapIndicesColumnCount;
+			if (rightIndex < resourceMapIndices.Length)
+			{
+				var indice = resourceMapIndices[rightIndex];
+				nearbyEnemyBase += indice.EnemyBaseCount;
+				nearbyEnemy += indice.EnemyUnitCount;
+				indiceCount++;
+			}
+
+			var downIndex = x + (y - 1) * resourceMapIndicesColumnCount;
+			if (downIndex >= 0)
+			{
+				var indice = resourceMapIndices[downIndex];
+				nearbyEnemyBase += indice.EnemyBaseCount;
+				nearbyEnemy += indice.EnemyUnitCount;
+				indiceCount++;
+			}
+
+			var upIndex = x + (y + 1) * resourceMapIndicesColumnCount;
+			if (upIndex < resourceMapIndices.Length)
+			{
+				var indice = resourceMapIndices[upIndex];
+				nearbyEnemyBase += indice.EnemyBaseCount;
+				nearbyEnemy += indice.EnemyUnitCount;
+				indiceCount++;
+			}
+
+			return (indiceCount, nearbyEnemy, nearbyEnemyBase);
+		}
+
+		public ResourceIndice GetIndice(int i)
+		{
+			if (resourceMapIndices == null || i >= resourceMapIndices.Length)
+				return null;
+
+			return resourceMapIndices[i];
+		}
+
+		public ResourceIndice FindClosestIndiceFromCPos(CPos cpos)
+		{
+			var maxDist = int.MaxValue;
+			var best = 0;
+
+			for (var i = 0; i < resourceMapIndices.Length; i++)
+			{
+				if ((resourceMapIndices[i].IndiceCenter - cpos).LengthSquared < maxDist)
+				{
+					maxDist = (resourceMapIndices[i].IndiceCenter - cpos).LengthSquared;
+					best = i;
+				}
```

Comment:
```suggestion
				var index = resourceMapIndices[i];
				var dist = (index.IndiceCenter - cpos).LengthSquared;
				if (dist < maxDist)
				{
					maxDist = dist;
					best = i;
				}
```

We can reuse calculations